libia2/include/ia2.h

```diff
@@ -264,13 +254,86 @@ static int insecure_pkey_mprotect(void *ptr, size_t len, int prot, int pkey) {
       *n_to_alloc = 0;                                                         \
     }                                                                          \
   }                                                                            \
-  char *ia2_stackptrs[n + 1] IA2_SHARED_DATA;                                  \
-  __attribute__((constructor)) static void init_stacks() {                     \
-    ensure_pkeys_allocated(&ia2_n_pkeys_to_alloc);                             \
+                                                                               \
+  /* one page of stack pointers for each compartment */                        \
+  char *ia2_stackptrs[PAGE_SIZE * (n + 1)] __attribute__((aligned(PAGE_SIZE))) \
+  IA2_SHARED_DATA;                                                             \
```

Comment:
Move `ia2_stackptrs` out of `ia2_shared_data` if we're going to protect it?

Reply:
My reasoning was that `ia2_stackptrs` is, in aggregate, shared, so there's no one compartment for which we want to mprotect it... but we do need to `pkey_mprotect` each page of it based on the compartment whose stack pointers it will hold.

Because we do `pkey_mprotect` each page of it, though, I guess it's also fine to not be `IA2_SHARED_DATA`--I tested it and it does work .

Reply:
yeah it's not a correctness thing, either should work. I was thinking that it might be confusing to debug later if we see something in ia2_shared_data getting pkey_mprotected

Reply:
This is now all obsolete as we use thread-local `ia2_stackptr_N` variables instead of an `ia2_stackptrs` array.